src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -52,107 +54,240 @@ final class P11SecretKeyFactory extends SecretKeyFactorySpi {
-        keyTypes.put(name, l);
-        keyTypes.put(name.toUpperCase(Locale.ENGLISH), l);
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA1AndAES_128",
+                CKP_PKCS5_PBKD2_HMAC_SHA1, 128));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA224AndAES_128",
+                CKP_PKCS5_PBKD2_HMAC_SHA224, 128));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA256AndAES_128",
+                CKP_PKCS5_PBKD2_HMAC_SHA256, 128));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA384AndAES_128",
+                CKP_PKCS5_PBKD2_HMAC_SHA384, 128));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA512AndAES_128",
+                CKP_PKCS5_PBKD2_HMAC_SHA512, 128));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA1AndAES_256",
+                CKP_PKCS5_PBKD2_HMAC_SHA1, 256));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA224AndAES_256",
+                CKP_PKCS5_PBKD2_HMAC_SHA224, 256));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA256AndAES_256",
+                CKP_PKCS5_PBKD2_HMAC_SHA256, 256));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA384AndAES_256",
+                CKP_PKCS5_PBKD2_HMAC_SHA384, 256));
+        putKeyInfo(new AESPBEKeyInfo("PBEWithHmacSHA512AndAES_256",
+                CKP_PKCS5_PBKD2_HMAC_SHA512, 256));
+
+        putKeyInfo(new PBKDF2KeyInfo("PBKDF2WithHmacSHA1",
+                CKP_PKCS5_PBKD2_HMAC_SHA1));
+        putKeyInfo(new PBKDF2KeyInfo("PBKDF2WithHmacSHA224",
+                CKP_PKCS5_PBKD2_HMAC_SHA224));
+        putKeyInfo(new PBKDF2KeyInfo("PBKDF2WithHmacSHA256",
+                CKP_PKCS5_PBKD2_HMAC_SHA256));
+        putKeyInfo(new PBKDF2KeyInfo("PBKDF2WithHmacSHA384",
+                CKP_PKCS5_PBKD2_HMAC_SHA384));
+        putKeyInfo(new PBKDF2KeyInfo("PBKDF2WithHmacSHA512",
+                CKP_PKCS5_PBKD2_HMAC_SHA512));
+
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA1",
+                CKM_PBA_SHA1_WITH_SHA1_HMAC, 160));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA224",
+                CKM_NSS_PKCS12_PBE_SHA224_HMAC_KEY_GEN, 224));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA256",
+                CKM_NSS_PKCS12_PBE_SHA256_HMAC_KEY_GEN, 256));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA384",
+                CKM_NSS_PKCS12_PBE_SHA384_HMAC_KEY_GEN, 384));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA512",
+                CKM_NSS_PKCS12_PBE_SHA512_HMAC_KEY_GEN, 512));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA512/224",
+                CKM_NSS_PKCS12_PBE_SHA512_HMAC_KEY_GEN, 512));
+        putKeyInfo(new P12MacPBEKeyInfo("HmacPBESHA512/256",
+                CKM_NSS_PKCS12_PBE_SHA512_HMAC_KEY_GEN, 512));
     }
 
-    // returns the PKCS11 key type of the specified algorithm
-    // no pseudo KeyTypes
     static long getPKCS11KeyType(String algorithm) {
         long kt = getKeyType(algorithm);
         if (kt == -1 || kt > PCKK_ANY) {
+            // Replace pseudo key type.
```

Comment:
Well, it seems clearer to have the "No pseudo key types" comment as method comment.

Reply:
Ok, moved as a method comment and replaced by the suggested wording.